### PR TITLE
[RFC] Clean up test/tridiag

### PR DESCRIPTION
**Pull request by kshyatt**
We had two loops doing very similar things. I tried to combine all the tests in a reasonable way. There's probably more that could be done here.

### CODE REVIEW

File: test/linalg/tridiag.jl [hunk 1]
```diff
@@ -228,118 +16,180 @@ end
...
+            end
+            # enable when deprecations for 0.7 are dropped
+            # @test_throws MethodError SymTridiagonal(dv, GenericArray(ev))
+            # @test_throws MethodError SymTridiagonal(GenericArray(dv), ev)
+            # @test_throws MethodError Tridiagonal(GenericArray(ev), dv, GenericArray(ev))
+            # @test_throws MethodError Tridiagonal(ev, GenericArray(dv), ev)
+        end
+        @testset "interconversion of Tridiagonal and SymTridiagonal" begin
+            @test Tridiagonal(dl, d, dl) == SymTridiagonal(d, dl)
+            @test SymTridiagonal(d, dl) == Tridiagonal(dl, d, dl)
```
Comment (fredrikekre) [inline test/linalg/tridiag.jl]: This test is completely equivalent with the one on the line above.

File: test/linalg/tridiag.jl [hunk 2]
```diff
@@ -228,118 +16,180 @@ end
...
-                @test_throws ArgumentError A[3, 2] = 1 # test assignment on the subdiagonal
-                @test_throws ArgumentError A[2, 3] = 1 # test assignment on the superdiagonal
-                @test_throws ArgumentError A[1, 3] = 1 # test assignment off the main/sub/super diagonal
+                @test_throws ArgumentError A[1, 3]   = 1 # test assignment off the main/sub/super diagonal
+                if mat_type == Tridiagonal
+                    @test (A[3, 3] = A[3, 3]; A == fA) # test assignment on the main diagonal
+                    @test (A[3, 2] = A[3, 2]; A == fA) # test assignment on the subdiagonal
+                    @test (A[2, 3] = A[2, 3]; A == fA) # test assignment on the superdiagonal
+                    @test ((A[1, 3] = 0) == 0; A == fA) # test zero assignment off the main/sub/super diagonal
+                else
```
Comment (fredrikekre) [inline test/linalg/tridiag.jl]: Add a comment `else # mat_type == SymTridiagonal` for clarity?

File: test/linalg/tridiag.jl [hunk 3]
```diff
@@ -228,118 +16,180 @@ end
...
+                    @test_throws ArgumentError A[2, 3] = 1 # test assignment on the superdiagonal
+                end
             end
             @testset "Diagonal extraction" begin
-                @test diag(A,1) == b
-                @test diag(A,-1) == b
-                @test diag(A,0) == a
-                @test diag(A,n-1) == zeros(elty,1)
-                @test_throws ArgumentError diag(A,n+1)
+                @test diag(A, 1) == (mat_type == Tridiagonal ? du : dl)
```
Comment (fredrikekre) [inline test/linalg/tridiag.jl]: This could test with `===`. and likewise the next two tests.

File: test/linalg/tridiag.jl [hunk 4]
```diff
@@ -228,118 +16,180 @@ end
...
+                    @test (A[3, 2] = A[3, 2]; A == fA) # test assignment on the subdiagonal
+                    @test (A[2, 3] = A[2, 3]; A == fA) # test assignment on the superdiagonal
+                    @test ((A[1, 3] = 0) == 0; A == fA) # test zero assignment off the main/sub/super diagonal
+                else
+                    @test ((A[3, 3] = A[3, 3]) == A[3, 3]; A == fA) # test assignment on the main diagonal
+                    @test_throws ArgumentError A[3, 2] = 1 # test assignment on the subdiagonal
+                    @test_throws ArgumentError A[2, 3] = 1 # test assignment on the superdiagonal
+                end
             end
             @testset "Diagonal extraction" begin
```
Comment (fredrikekre) [inline test/linalg/tridiag.jl]: Could also be worth adding a test for `diag(A)`, i.e. without specifying a diagonal.

File: test/linalg/tridiag.jl [hunk 5]
```diff
@@ -228,118 +16,180 @@ end
...
-                    @test isa(round.(Int,A), SymTridiagonal)
-                    @test trunc.(Int,A) == trunc.(Int,fA)
-                    @test isa(trunc.(Int,A), SymTridiagonal)
-                    @test ceil.(Int,A) == ceil.(Int,fA)
-                    @test isa(ceil.(Int,A), SymTridiagonal)
-                    @test floor.(Int,A) == floor.(Int,fA)
-                    @test isa(floor.(Int,A), SymTridiagonal)
+                for f in (round, trunc, floor, ceil)
+                    fds = [f.(d) for d in ds]
+                    @test f.(A) == mat_type(fds...)
```
Comment (fredrikekre) [inline test/linalg/tridiag.jl]: Perhaps remove the `isa` tests here and do `@test (f.(A))::mat_type == mat_type(fds...)` for even more code deletion.

File: test/linalg/tridiag.jl [hunk 6]
```diff
@@ -348,109 +198,114 @@ let n = 12 #Size of matrix problem to test
...
-                @testset "Rounding to Ints" begin
-                    @test round.(Int,A) == round.(Int,fA)
-                    @test isa(round.(Int,A), Tridiagonal)
-                    @test trunc.(Int,A) == trunc.(Int,fA)
-                    @test isa(trunc.(Int,A), Tridiagonal)
-                    @test ceil.(Int,A) == ceil.(Int,fA)
-                    @test isa(ceil.(Int,A), Tridiagonal)
-                    @test floor.(Int,A) == floor.(Int,fA)
-                    @test isa(floor.(Int,A), Tridiagonal)
+                if elty <: Real
```
Comment (fredrikekre) [inline test/linalg/tridiag.jl]: Seems like these two `if`s could be collapsed to one level with `if elty <: LinAlg.BlasReal`

**Review by fredrikekre — COMMENTED** [review]
Looks nicer, some (optional) comments :sweden:

**Comment by kshyatt** [comment]
Thanks for the detailed review, @fredrikekre. These are great suggestions <3!

File: test/linalg/tridiag.jl (hunk 1 above)
Comment (kshyatt) [reply test/linalg/tridiag.jl]: A bit of a theme in this file...

**Review by fredrikekre — APPROVED** [review]
Nice!

**Pull request merged**